0000-stream/0000-stream.md

```diff
@@ -0,0 +1,174 @@
+
+## Specification
+
+All STREAM packets are intended to be sent in the `data` of an [ILP Prepare, Fulfill, or Reject](../0027-interledger-protocol-4/0027-interledger-protocol-4.md) packet.
+
+### Encryption
+
+All STREAM packets are encrypted using [AES-256-GCM](https://en.wikipedia.org/wiki/Galois/Counter_Mode) with a random 12-byte Initialization Vector (IV) and a 16-byte Authentication Tag.
+
+#### Encryption Envelope
+
+See the [ASN.1 definition](../asn1/Stream.asn) for the formal specification.
+
+| Field | Type | Description |
+|---|---|---|
+| Random IV  | 12-Byte UInt | Nonce used as input to the AES-GCM algorithm. Also ensures conditions are random. Senders MUST NOT encrypt two packets with the same nonce |
+| Authentication Tag   | 16-Byte UInt | Authentication tag produced by AES-GCM encryption that ensures data integrity |
+| Ciphertext | 0-32739 Bytes | Encrypted data (see below for contents) |
+
+#### Encryption Pseudocode
+
+Note that the encryption key is derived from the shared secret using an HMAC.
+
+```
+iv = random_bytes(12)
+encryption_key = hmac_sha256(shared_secret, "ilp_stream_encryption")
+{ ciphertext, auth_tag } = aes_256_gcm(encryption_key, iv, data)
+```
+
+### STREAM Packet
+
+See the [ASN.1 definition](../asn1/Stream.asn) for the formal specification.
+
+| Field | Type | Description |
+|---|---|---|
+| Version | UInt8 | `1` for this version |
+| ILP Packet Type | UInt8 | ILPv4 packet type this STREAM packet MUST be sent in (`12` for Prepare, `13` for Fulfill, and `14` for Reject). A sender or receiver MUST discard a STREAM packet that comes in on the wrong ILP Packet Type. |
+| Sequence | VarUInt | Identifier for this request/response. A STREAM packet sent with an ILP Fulfill or Reject MUST have the same `Sequence` as the ILP Prepare packet it is a response to. A sender MUST discard a STREAM packet in which the `Sequence` does not match the STREAM packet sent with their ILP Prepare. |
+| Prepare Amount | VarUInt | If the STREAM packet is sent on an ILP Prepare, this represents the minimum the receiver should accept. If the packet is sent on an ILP Fulfill or Reject, this represents the amount that the receiver got in the Prepare. |
+| Frames | SEQUENCE OF Frame | The rest of the packet is comprised of type- and length-prefixed Frames, as specified below. Note that the array of frames is NOT length-prefixed. |
+
+**TODO**: Should we length-prefix the frames array and make the padding a field in the packet format to better comply with OER? Note that we cannot length-prefix the padding because in certain edge cases it could make it impossible to pad to the right length when the length prefix is included.
```

Comment:
I think it's better to length prefix the frames array than use 0x00 as a delimiter. Then if we want to extend the protocol later we wouldn't have a permanent 0 byte delimiter as support for previous versions

Reply:
Actually I tried this an ran into the problem that length-prefixing the padding potentially changes the amount of padding you need. I think it's possible to have a situation where it is impossible to fill the packet exactly while putting the correct length prefix. I think this is why QUIC implemented the padding this way as well.

> Then if we want to extend the protocol later we wouldn't have a permanent 0 byte delimiter as support for previous versions

What do you mean by this? It's easy to add new frames to the protocol because they'll just be ignored by old versions. I think that's a better way to extend the protocol than modifying the packet header.